README.md
worddimensions
==============

a pattern-matching game based on words


NEWS
55

* New translations

53

* Fix license text
* New translations

51

* Add support for screen rotation
* Make separate games for shapes, numbers, and words

49

* New translations

47

* New translations

45

* New translations

43

BUG FIX:
* Fixed name change in radio button (#4028)

41

ENHANCEMENTS:
* GTK-3 version
* Added summary to activity.info
* Added support for touch

39

BUG FIX:
* Closing the Welcome message clicks on a card too (#3567)
* Sometimes cards disappear (#3568)
* Detect and correct race condition (#3579)
(Thanks to Manuel Kaufmann for help tracking down these bugs)

38

ENHANCEMENTS:
* New translations
* Add stop button to help animation

BUG FIX:
http://dev.laptop.org/ticket/11804
* Refactoring of animation to use less memory
* Refactoring of deck to use less memory

37

BUG FIX:
* Fixed typo that prevented help animation from terminating

ENHANCEMENT:
* Autoadvance levels when play improves

36

ENHANCEMENTS:
* Support for dragging cards
* Visual feedback on mismatches
* Cursor when editing cards
* New help animation on first load
* Copy scores to clipboard (for import into Simple Graph)
* Notification when loading new games
* Slight delay after match before dealing new cards to allow for some
  scrutiny of results
* Addition of toolbar help menus
* Rebranding as Dimensions (a more apt name)
* Updated wiki page (See http://wiki.sugarlabs.org/go/Activities/Dimensions)
* New translations
(Tip of the hat to Mark Battley for great feedback and suggestions.)

35

ENHANCEMENTS:
* New translations

34

ENHANCEMENTS:
* New translations

BUG FIX:
* Fixed problem with animated cards missing destinations

33

ENHANCEMENTS:
* New translations
* New end-of-game graphics

32

ENHANCEMENT:
* Using Cairo graphics
* Animate loading of new cards

BUG FIX:
* Set robot time during "attract mode"

31

ENHANCEMENT:
* Animate matches when moving to discard pile

BUG FIX:
* FIx GPLv3 license address

30

ENHANCEMENTS:
* Switching buttons to radio buttons
* More button icon refactoring

29

ENHANCEMENT:
* "Standardized" icons

28

ENHANCEMENT:
* Added "attract mode" on first launch

27

ENHANCEMENTS:
* Smiley-face "reward" for finishing a game
* Improved workflow for importing custom cards

BUG FIXES:
* Work around datastore.find issue on 0.84 builds
* Work around chooser issues with old Sugar builds (#2398)

26

ENHANCEMENT:
* Import custom cards from Journal

25

ENHANCEMENTS:
* Quipu number game
* Optimization of sprite refresh

24

ENHANCEMENTS:
* Added beginner level
* Detect unsolvable game-play situations

23

BUG FIXES:
* Byron Corrales fixed the robot/word-edit conflict (#2057)
* General code cleanup

22

BUG FIX:
* Fixed SVG bug in old Sugar builds

21

BUG FIXES:
* general clean up of sprite library
* updated licenses

20

ENHANCEMENT:
* Added more combinations to the dice game, e.g., 5=4+1, 5=3+2

BUG FIXES:
* Fixed bug in sprite library
* Using refactored card generator

19

BUG FIXES:
* Fixed bug with Mayan 33
* Cleaned up some icons that had SVG problems on old Sugar builds

18

ENHANCEMENTS:
* Word-game lists can be edited
* Mayan number game
* "Fileless" SVG generation

BUG FIX:
* Fixed SVG problem on old Sugar builds

17

BUG FIXES:
* Fixed Class problem with command-line version
* Cleaned up license language (GPL v3)

16

ENHANCEMENTS:
* Simple sharing/collaborating
* Save/resume from Journal 
* New (improved?) toolbar layout
* Match-history animation at end game

BUG FIX:
* Fixed bug preventing play on old Sugar instances

15

ENHANCEMENTS:
* Added low/high levels for game play
* New activity icon

14

ENHANCEMENT:
* Computer assistant

13

ENHANCEMENT:
* More number games

12

ENHANCEMENTS:
* Reverting to three colors to simplify game play
* Experimenting with different number-game card styles
* Adding bold and italic to word cards
* Consolidate grid after dealing extra cards
* Refactoring of sprites library

11

BUG FIXES:
* Add check for match on first deal
* Fixed bug in extra card positions

10

ENHANCEMENTS:
* es, fr translation
* Rotated cards to enable larger/longer text strings

BUG FIX:
* Cleaned up match logic for word games

9

ENHANCEMENTS:
* Auto-deal of extra cards when needed
* Added glue for launching outside of Sugar
* Save "low score" to Journal
* Added numbers and words games
* Autogenerate cards on the fly
* Minutes:seconds display for total time
* General code cleanup

8

ENHANCEMENT:
* Added unselect

BUG FIX:
* Fixed bug in new game labels

7

ENHANCEMENTS:
* More distinct colors on cards
* Count seconds between matches
* Detect game over
* Automatically deals extra cards when necessary

6

BUG FIX:
* Fixed card selection bug

5

ENHANCEMENT:
* Added search function

4

ENHANCEMENT:
* Add 3 cards when you get stuck

3

ENHANCEMENTS:
* Eliminate naming alert from pre-0.86 toolbar
* New .pot file

2

ENHANCEMENT:
* New icon

BUG FIX:
* Fixed problem with the New Game button

1

* Version 1 of a visual matching game. The object is to find sets of
  three cards where each attribute—color, shape, number of elements,
  and shading—either match on all three cards or are different on all
  three cards. Version 1 doesn't yet support sharing with multiple
  players or saving to the Journal, but it can be played by a single
  player.


dimensions.py
#!/usr/bin/env python

#Copyright (c) 2009-12 Walter Bender

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the
# Free Software Foundation, Inc., 59 Temple Place - Suite 330,
# Boston, MA 02111-1307, USA.

import gi
from gi.repository import Gtk
from gi.repository import Gdk

from gettext import gettext as _
import os

from game import Game
from constants import PRODUCT, HASH, ROMAN, WORD, CHINESE, MAYAN, DICE, DOTS, \
                      STAR, LINES, INCAN


class DimensionsMain:

    def __init__(self):
        # create a new window
        self.win = Gtk.Window(type=Gtk.WindowType.TOPLEVEL)
        self.win.maximize()
        self.win.set_title("%s: %s" % (_("Dimensions"),
                           _("Click on cards to create sets of three.")))
        self.win.connect("delete_event", lambda w, e: Gtk.main_quit())

        menu0 = Gtk.Menu()
        menu_items = Gtk.MenuItem(_("beginner"))
        menu0.append(menu_items)
        menu_items.connect("activate", self._level_cb, 0)
        menu_items.show()
        menu_items = Gtk.MenuItem(_("intermediate"))
        menu0.append(menu_items)
        menu_items.connect("activate", self._level_cb, 1)
        menu_items.show()
        menu_items = Gtk.MenuItem(_("expert"))
        menu0.append(menu_items)
        menu_items.connect("activate", self._level_cb, 2)
        menu_items.show()
        #TRANS: Level of difficulty
        level_menu = Gtk.MenuItem(_("Level"))
        level_menu.show()
        level_menu.set_submenu(menu0)

        menu1 = Gtk.Menu()
        '''
        menu_items = Gtk.MenuItem(_("New pattern game"))
        menu1.append(menu_items)
        menu_items.connect("activate", self._new_game_cb, 'pattern')
        menu_items.show()
        menu_items = Gtk.MenuItem(_("New number game"))
        menu1.append(menu_items)
        menu_items.connect("activate", self._new_game_cb, 'number')
        menu_items.show()
        '''
        menu_items = Gtk.MenuItem(_("New word game"))
        menu1.append(menu_items)
        menu_items.connect("activate", self._new_game_cb, 'word')
        menu_items.show()
        game_menu = Gtk.MenuItem(_("Game"))
        game_menu.show()
        game_menu.set_submenu(menu1)

        menu2 = Gtk.Menu()
        menu_items = Gtk.MenuItem(_("Robot on/off"))
        menu2.append(menu_items)
        menu_items.connect("activate", self._robot_cb)
        menu_items.show()
        menu_items = Gtk.MenuItem(_("5 minutes"))
        menu2.append(menu_items)
        menu_items.connect("activate", self._robot_time_cb, 300)
        menu_items.show()
        menu_items = Gtk.MenuItem(_("2 minutes"))
        menu2.append(menu_items)
        menu_items.connect("activate", self._robot_time_cb, 120)
        menu_items.show()
        menu_items = Gtk.MenuItem(_("1 minute"))
        menu2.append(menu_items)
        menu_items.connect("activate", self._robot_time_cb, 60)
        menu_items.show()
        menu_items = Gtk.MenuItem(_("30 seconds"))
        menu2.append(menu_items)
        menu_items.connect("activate", self._robot_time_cb, 30)
        menu_items.show()
        menu_items = Gtk.MenuItem(_("15 seconds"))
        menu2.append(menu_items)
        menu_items.connect("activate", self._robot_time_cb, 15)
        menu_items.show()
        tool_menu = Gtk.MenuItem(_('Play with the computer'))
        tool_menu.show()
        tool_menu.set_submenu(menu2)

        '''
        menu3 = Gtk.Menu()
        menu_items = Gtk.MenuItem(_("Product"))
        menu3.append(menu_items)
        menu_items.connect("activate", self._number_card_O_cb, PRODUCT)
        menu_items.show()
        menu_items = Gtk.MenuItem(_("Roman"))
        menu3.append(menu_items)
        menu_items.connect("activate", self._number_card_O_cb, ROMAN)
        menu_items.show()
        menu_items = Gtk.MenuItem(_("Word"))
        menu3.append(menu_items)
        menu_items.connect("activate", self._number_card_O_cb, WORD)
        menu_items.show()
        menu_items = Gtk.MenuItem(_("Chinese"))
        menu3.append(menu_items)
        menu_items.connect("activate", self._number_card_O_cb, CHINESE)
        menu_items.show()
        menu_items = Gtk.MenuItem(_("Mayan"))
        menu3.append(menu_items)
        menu_items.connect("activate", self._number_card_O_cb, MAYAN)
        menu_items.show()
        menu_items = Gtk.MenuItem(_("Quipu"))
        menu3.append(menu_items)
        menu_items.connect("activate", self._number_card_O_cb, INCAN)
        menu_items.show()
        menu_items = Gtk.MenuItem(_("Hash"))
        menu3.append(menu_items)
        menu_items.connect("activate", self._number_card_C_cb, HASH)
        menu_items.show()
        menu_items = Gtk.MenuItem(_("Dice"))
        menu3.append(menu_items)
        menu_items.connect("activate", self._number_card_C_cb, DICE)
        menu_items.show()
        menu_items = Gtk.MenuItem(_("Dots"))
        menu3.append(menu_items)
        menu_items.connect("activate", self._number_card_C_cb, DOTS)
        menu_items.show()
        menu_items = Gtk.MenuItem(_("Star"))
        menu3.append(menu_items)
        menu_items.connect("activate", self._number_card_C_cb, STAR)
        menu_items.show()
        menu_items = Gtk.MenuItem(_("Lines"))
        menu3.append(menu_items)
        menu_items.connect("activate", self._number_card_C_cb, LINES)
        menu_items.show()
        num_menu = Gtk.MenuItem("Numbers")
        num_menu.show()
        num_menu.set_submenu(menu3)
        '''

        # A vbox to put a menu and the canvas in:
        vbox = Gtk.VBox(False, 0)
        self.win.add(vbox)
        vbox.show()

        menu_bar = Gtk.MenuBar()
        vbox.pack_start(menu_bar, False, False, 2)
        menu_bar.show()

        canvas = Gtk.DrawingArea()
        vbox.pack_end(canvas, True, True, 0)
        canvas.show()

        menu_bar.append(game_menu)
        menu_bar.append(level_menu)
        menu_bar.append(tool_menu)
        # menu_bar.append(num_menu)
        self.win.show_all()

        # Join the activity
        self.vmw = Game(canvas)
        self.vmw.win = self.win
        self.vmw.activity = self
        self.vmw.card_type = 'word'
        self.vmw.level = 1
        self.vmw.robot = False
        self.vmw.robot_time = 60
        self.vmw.numberO = PRODUCT
        self.vmw.numberC = HASH
        self.vmw.matches = 0
        self.vmw.robot_matches = 0
        self.load_score()
        self.vmw.word_lists = [[_('mouse'), _('cat'), _('dog')],\
                           [_('cheese'), _('apple'), _('bread')],\
                           [_('moon'), _('sun'), _('earth')]]

        self.vmw.new_game()

    def load_score(self):
        try:
            f = file(os.path.join(os.path.abspath('.'),
                                  'visualmatch.score'), "r")
            s = f.readlines()
            f.close
            self.vmw.low_score = [int(s[0].split(':')[1].strip()),
                                  int(s[1].split(':')[1].strip())]
            print "low score is: %s" % (self.vmw.low_score)
        except:
            self.vmw.low_score = [-1, -1]

    def save_score(self):
        f = file(os.path.join(os.path.abspath('.'), 'visualmatch.score'), "w")
        f.writelines(["low_score_beginner:%d\n" % int(self.vmw.low_score[0]),
                      "low_score_expert:%d\n" % int(self.vmw.low_score[1])])
        f.close

    def set_title(self, title):
        self.win.set_title(title)

    def _new_game_cb(self, widget, game):
        self.vmw.card_type = game
        self.vmw.new_game()
        return True

    def _level_cb(self, widget, level):
        self.vmw.level = level
        self.vmw.new_game()

    def _robot_cb(self, widget):
        if self.vmw.robot:
            self.vmw.robot = False
        else:
            self.vmw.robot = True

    def _robot_time_cb(self, widget, time):
        self.vmw.robot_time = time

    def _number_card_O_cb(self, widget, numberO):
        self.vmw.numberO = numberO
        self.vmw.card_type = 'number'
        self.vmw.new_game()

    def _number_card_C_cb(self, widget, numberC):
        self.vmw.numberC = numberC
        self.vmw.card_type = 'number'
        self.vmw.new_game()


def main():
    Gtk.main()
    return 0


if __name__ == "__main__":
    DimensionsMain()
    main()
